### PR TITLE
Tidy up early bailout logic at start of begin phase

### PR DESCRIPTION
**Pull request by acdlite**
An early bailout is when there are no scheduled changes to props, state, or context, so we bail out at the start of the begin phase without running the component body.

The logic in this path is getting hard to follow, so I refactored it a bit. There shouldn't be any behavioral change.

I suggest reviewing each commit separately — the first one looks big but that's only because I extracted a bunch of lines into a separate function. The second commit is slightly more interesting.

As a bonus, this should make bailouts slightly faster because it moves `checkIfContextChanged` out of the common path.

Ideally this early bailout path would happen even earlier, before the begin phase. Perhaps during reconciliation of the parent fiber's children. That would require a much bigger refactor, but something to consider for the future.

### CODE REVIEW

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js [hunk 1]
```diff
@@ -3222,6 +3216,27 @@ function remountFiber(
...
 }
 
+function checkScheduledUpdateOrContext(
+  current: Fiber,
+  renderLanes: Lanes,
+): boolean {
+  // Before performing an early bailout, we must check if there are pending
+  // updates or context.
+  const updateLanes = current.lanes;
+  if (includesSomeLane(renderLanes, updateLanes)) {
```
Comment (rickhanlonii) [inline packages/react-reconciler/src/ReactFiberBeginWork.new.js:3226]: `current` is first in the function definition, but second in this invocation. snip snap, snap snip

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js [hunk 2]
```diff
@@ -3480,14 +3482,23 @@ function beginWork(
...
-      return attemptEarlyBailoutIfNoScheduledUpdate(
-        current,
-        workInProgress,
-        renderLanes,
-      );
     } else {
+      // Neither props nor legacy context changes. Check if there's a pending
+      // update or context change.
+      if (
+        !checkScheduledUpdateOrContext(current, renderLanes) &&
```
Comment (rickhanlonii) [inline packages/react-reconciler/src/ReactFiberBeginWork.new.js:3489]: Nit: this would be easier for me to read if it started with `has`

"not check" is harder to skim than "not has"

**Review by rickhanlonii — APPROVED** [review]
LGTM, a couple non-blocking nits

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js [hunk 3]
```diff
@@ -762,7 +758,7 @@ function updateCacheComponent(
     pushCacheProvider(workInProgress, freshCache);
   } else {
     // Check for updates
-    if (includesSomeLane(renderLanes, updateLanes)) {
+    if (includesSomeLane(current.lanes, renderLanes)) {
```
[the remark this answers is not in the document]
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js:761]: The order for this function doesn't really matter because the implementation is `A & B !== 0`, and we're not super consistent elsewhere, but to explain my thinking, I usually try to put them in  "subject-object" order, like English. So  this would be: "check if the update lanes include the render lanes." Although you could argue either one is the "subject", but for whatever reason that's how I think of it. I guess because in a typical scenario the update lanes are a larger set than the render lanes.

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js (hunk 1 above)
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js:3226]: Yeah based on my "subject-object" rationale this should be flipped

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js (hunk 2 above)
Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js:3489]: Yeah I suppose, I did it like that for consistency with the existing `checkIfWorkInProgressReceivedUpdate` function, which I think was originally named like that to avoid a clash with the variable it returned... which was named `hasWorkInProgressReceivedUpdate`

Comment (acdlite) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js:3489]: How about I'll assign to a named variable before negating it :D

Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js]: sgtm haha

File: packages/react-reconciler/src/ReactFiberBeginWork.new.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (rickhanlonii) [reply packages/react-reconciler/src/ReactFiberBeginWork.new.js:765]: Makes sense!
